//Define the used version of proto
syntax = "proto3";

package simulation;

message InitRequest {
    string StartDate      = 1 ;
    string EndDate        = 2 ;
    float DensityPerc     = 3 ;
}

message StepRequest {
  int32 numSteps = 1;
  map<string, bool> cell_state = 2;
}

message Position {
    int32 x = 1;
    int32 y = 2;
}

message StateResponse {
  int32 currentStep = 1;
  map<string, float> emissions = 2;
  map<string, int32> vehicles = 3;
  map<string, int32> state = 4;
  map<string, Position> cell_map = 5;
  bool hasEnded = 6;
}

//Define the service
service Simulator{
  rpc start_simulation(InitRequest) returns (StateResponse);
  rpc step(StepRequest) returns (StateResponse);
}
